### PR TITLE
first push

### PR DESCRIPTION
**Pull request by Daviderikmollberg**
First push. Know things that I will be fixing later are:

Add the subword script to samromur-asr corpus
Add Morfessor to the script
Add Unigram to the script
Change the run script to handle pretrained language models
Move from Sirlm to kenlm
Tidy up the LM step

### CODE REVIEW

File: s5_subwords/Readme.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+
+	<wav-filename>	<recording-info>	<recording-info>	<gender>	<age>	<prompt (spoken text)>	<utterance length>	vorbis	16000	1	Vorbis
+
+
+The `malromur_prep_data.sh` script divides the generated data:
+ 
+	local/malromur_prep_data.sh <path-to-audio-files> wav_info.txt data/all
+	utils/subset_data_dir_tr_cv.sh --cv-utt-percent 10 data/{all,training_data,test_data}
+
```
Comment (ingarun) [inline s5_subwords/Readme.md:37]: I find it better to have a set test set, for comparison later. Won't you get a different test set each time this command is run?

File: s5_subwords/Readme.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+## Running the scripts
+The run.sh script takes care of the training process. Three subword segmentation methods can be provided as command line argument with the run.sh script. 
+
+`./run.sh bpe` or
+`./run.sh kvistur` or
+`./run.sh unigram`
+
+If no argument is given, the BPE algorithm is applied by default. The model training architecture is described in [2].
+The last training step of the script (neural network training) might fail for some reasons; we have not used this step in our experiments.
```
Comment (ingarun) [inline s5_subwords/Readme.md:58]: Why not just leave out the neural network part until it is tested? Or say: Neural network training is not yet implemented or that it will be added soon?

File: s5_subwords/runV3.sh [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+# Audio data paths 
+AUDIO=/data/asr/malromur/malromur2017/correct
+
+#samromur_root=/data/asr/samromur/samromur_ldc
+METADATA=/home/derik/work/tools/normalize/malromur/normalized_files/malromur_metadata.tsv
+
+# Text corpus for the LM
+text_corpus=/home/derik/work/tools/normalize/malromur/normalized_files/corpus.txt
+
```
Comment (ingarun) [inline s5_subwords/runV3.sh:58]: A dependency check would be ok here. E.g. check whether srilm is installed and the directory and two files are ok. But not necessary since instructions in readme tell the user to install it.

File: s5_subwords/runV3.sh [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                              data/$lang/local/lm \
+                              6               
+
+  utils/format_lm.sh  data/$lang/lang \
+                      data/$lang/local/lm/lm.gz \
+                      data/$lang/local/dict/lexicon.txt \
+                      data/$lang/lang_test
+fi
+
+exit
```
Comment (ingarun) [inline s5_subwords/runV3.sh]: This "exit" needs to go away

File: s5_subwords/local/prep_metadata.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    utt2spk.append(f"{utt_id} {f[2]}\n")
+    
+    #This line will cause an error in versions of Samrómur where the speaker is unknown
+    if f[3][0] == 'f':
+        gender = 'f'
+    elif f[3][0] == 'm':
+        gender = 'm'
+    else:
+        gender = 'f'
+
```
Comment (ingarun) [inline s5_subwords/local/prep_metadata.py:39]: Out of curiosity, why do you assign f to unknown gender segments?

File: s5_subwords/local/prep_metadata.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        description="\nUsage: $0 <path-to-audio> <audio-meatadata-file> <create_kaldi_id's> \
+        This script will output the files text, wav.scp, utt2spk, spk2utt and spk2gender \
+        to data/all, data/train data/test with test/train/eval splits defined in the metadatafile. \
+        The script also makes a tokens file which which contains the lexicon in a the corpus \
+        Eg. samromur_prep_data.py /data/corpora/samromur/audio/ metadata.tsv\n")
+        
+    parser.add_argument("--audio", required=True, type=str,
+                        help="Path to audio corpus")
+    parser.add_argument("--metadata", required=True, type=str,
+                        help="Paht to the metadatafile")
```
Comment (ingarun) [inline s5_subwords/local/prep_metadata.py]: Paht -> Path

File: s5_subwords/runV3.sh [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                       data/$lang/$i \
+                       exp/$lang/mfcc/log/make_mfcc \
+                       exp/$lang/mfcc/$i || error 1 "Failed creating MFCC features";
+
+    steps/compute_cmvn_stats.sh data/$lang/$i \
+                                exp/$lang/mfcc/log/cmvn_stats \
+                                exp/$lang/mfcc/$i || exit 1;
+  done
+fi
+
```
Comment (ingarun) [inline s5_subwords/runV3.sh:128]: I would validate and fix the data dir after creating the mfccs

File: s5_subwords/runV3.sh [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+                    exp/$lang/tri1 \
+                    exp/$lang/tri1_ali || exit 1;
+
+  steps/train_lda_mllt.sh --cmd "$train_cmd" \
+                          4000 50000 \
+                          data/$lang/training \
+                          data/$lang/lang \
+                          exp/$lang/tri1_ali \
+                          exp/$lang/tri2b || exit 1;
+fi
```
Comment (ingarun) [inline s5_subwords/runV3.sh:182]: What does the b stand for in tri2b?

File: s5_subwords/runV3.sh [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+if [ $stage -le 8 ] && $decode_gmm; then
+  echo ============================================================================
+  echo "          Decoding tri3                          "
+  echo ============================================================================
+  tri=tri3b
+  utils/mkgraph.sh data/$lang/lang_test \
+                   exp/$lang/$tri \
+                   exp/$lang/$tri/graph || exit 1;
+
+  steps/decode.sh --nj $num_decode_jobs \
```
Comment (ingarun) [inline s5_subwords/runV3.sh]: Copy score.sh from somewhere into local

File: s5_subwords/local/prepare_dict_subword.sh [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+training=$2
+dir=$3
+
+set -e
+
+stage=0
+. ./cmd.sh
+. ./path.sh
+. ./utils/parse_options.sh || exit 1;
+
```
Comment (ingarun) [inline s5_subwords/local/prepare_dict_subword.sh:24]: In the Kaldi scripts the order of lines is usually different, so that configuration comes first, then path.sh and those, then usage and finally input files from the command line. It usually doesn't matter but I find it rather nice to always have it in the same order

File: s5_subwords/local/prepare_dict_subword.sh [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+# cut -d' ' -f2- $dir/lexicon.txt | tr ' ' '\n' | LC_ALL=C sort -u | sed '/^$/d' > $dir/nonsilence_phones.txt
+
+echo UNK >> $dir/nonsilence_phones.txt
+echo SIL > $dir/silence_phones.txt
+echo SIL >$dir/optional_silence.txt
+echo -n "" >$dir/extra_questions.txt
+
+glossaries="<UNK> <sil>"
+
+if [ $stage -le 0 ]; then
```
Comment (ingarun) [inline s5_subwords/local/prepare_dict_subword.sh]: stage = 0 again?

File: s5_subwords/local/subword_methods/bpe.sh [hunk 12]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash 
+#
+#
+# 2020 David Erik Mollberg 
+#
+
+lang=$1
+text_corpus=$2 
+num_merges=$3
+stage=2
```
Comment (ingarun) [inline s5_subwords/local/subword_methods/bpe.sh]: stage is not used and the comments could be in English

File: s5_subwords/local/subword_methods/train_morfessor.sh [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+dir=$2
+testdata=$3
+
+mkdir -p $dir
+cut -d' ' -f2- $corpus | sed 's/ /\n/g' > $dir/corpus 
+
+corpus=$dir/corpus
+
+# Create 1M substring seed lexicon direct from a pretokenized corpus
+freq_substr.py --lex-size 1000000 < $corpus > $dir/freq_substr.1M
```
Comment (ingarun) [inline s5_subwords/local/subword_methods/train_morfessor.sh]: I don't find this python file or where the morfessor executables come from

**Review by ingarun — CHANGES_REQUESTED** [review]
I have written some comments and have some questions. Everything ran with the BPE method, after I removed the "exit" and added score.sh, but I got a WER of 82% so I will have to figure out why with you

File: s5_subwords/Readme.md (hunk 2 above)
Comment (Daviderikmollberg) [reply s5_subwords/Readme.md:58]: I have it almost ready, I removed it for this pull request and will focus on it later

File: s5_subwords/local/prep_metadata.py (hunk 5 above)
Comment (Daviderikmollberg) [reply s5_subwords/local/prep_metadata.py:39]: No particular reason, I think I saw it done somewhere.

File: s5_subwords/local/subword_methods/train_morfessor.sh (hunk 13 above)
Comment (Daviderikmollberg) [reply s5_subwords/local/subword_methods/train_morfessor.sh]: Ah, that wasn't supposed to go with this commit. I will be adding the method later.

File: s5_subwords/local/prep_metadata.py (hunk 5 above)
Comment (judyfong) [reply s5_subwords/local/prep_metadata.py:39]: isn't this done in the althingi asr scripts?

File: s5_subwords/Readme.md (hunk 1 above)
Comment (Daviderikmollberg) [reply s5_subwords/Readme.md:37]: Good point, will make a note of this.

File: s5_subwords/runV3.sh (hunk 8 above)
Comment (Daviderikmollberg) [reply s5_subwords/runV3.sh:182]: Good question, not sure

**Comment by Daviderikmollberg** [comment]
Addressed all the comments and found the bug which was causing the high WER. The WER is now 12%
